tests/operator_test.go

```diff
@@ -206,4 +209,91 @@ var _ = Describe("Operator delete CDI tests", func() {
 		err = f.CdiClient.CdiV1alpha1().CDIs().Delete(cr.Name, &metav1.DeleteOptions{DryRun: []string{"All"}})
 		Expect(err).ToNot(HaveOccurred())
 	})
+
+	It("should deploy components that tolerate CriticalAddonsOnly taint", func() {
+		var cdiPods *corev1.PodList
+		var err error
+		By("finding all nodes that are running cdi components", func() {
+			cdiPods, err = f.K8sClient.CoreV1().
+				Pods(f.CdiInstallNs).List(metav1.ListOptions{})
+			Expect(err).ShouldNot(HaveOccurred(), "failed listing cdi pods")
+			Expect(len(cdiPods.Items)).
+				To(BeNumerically(">", 0), "no cdi pods found")
+		})
+
+		// nodes that run cdi components.
+		cdiNodes := make(map[string]*corev1.Node)
+		By("adding taints to any node that runs cdi component", func() {
+			criticalPodTaint := corev1.Taint{
+				Key:    "CriticalAddonsOnly",
+				Value:  "",
+				Effect: corev1.TaintEffectNoExecute,
+			}
+
+			for _, cdiPod := range cdiPods.Items {
+				cdiNode, err := f.K8sClient.CoreV1().Nodes().Get(cdiPod.Spec.NodeName, metav1.GetOptions{})
+				Expect(err).ShouldNot(HaveOccurred(), "failed retrieving node")
+				hasTaint := false
+				// check if node already has the taint
+				for _, taint := range cdiNode.Spec.Taints {
+					if reflect.DeepEqual(taint, criticalPodTaint) {
+						// node already have the taint set
+						hasTaint = true
+						break
+					}
+				}
+				if hasTaint {
+					continue
+				}
+				cdiNode.ResourceVersion = ""
+				cdiNodes[cdiPod.Spec.NodeName] = cdiNode
+				cdiNodeCopy := cdiNode.DeepCopy()
+				cdiNodeCopy.Spec.Taints = append(cdiNodeCopy.Spec.Taints, criticalPodTaint)
+				_, err = f.K8sClient.CoreV1().Nodes().Update(cdiNodeCopy)
+				Expect(err).ShouldNot(HaveOccurred(), "failed setting taint on node")
+			}
+		})
+
+		defer func() {
+			var errors []error
+			By("restoring nodes")
+			for _, nodeSpec := range cdiNodes {
+				_, err := f.K8sClient.CoreV1().Nodes().Update(nodeSpec)
+				if err != nil {
+					errors = append(errors, err)
+				}
+			}
+			Expect(errors).Should(BeEmpty(), "failed restoring one or more nodes")
+		}()
+
+		By("watching for terminated cdi pods", func() {
```

Comment:
Shouldn't we technically be watching before the taint is added?  Is there not a possibility of missing an event otherwise?

Reply:
Technically, it is possible... since it takes the pod controller some time (seconds) to actually delete pods after they were marked with the deletion timestamp, we catch them.

But I agree with you, it's better to start the watching routine before we taint the nodes... I've updated the pr. Thanks :smile: